### ISSUE TITLE
How to use ListFilter?

### ISSUE THREAD

**Issue by marnickmenting** [vdhicts/cyberfusion-cluster-api-client#46]
## Goal

I am trying to filter certificates on common_names. In Postman I do this:

GET {{url}}/certificates?filter=common_names:["urliwanttoknowstatusof.nl"]

This works. But how do I put this in ListFilter. I do the following:

```
$filter = new ListFilter();
$filter->setFilter(['common_names:["urliwanttoknowstatusof.nl"]']);
$cert = $cf->api->certificates()->list($filter)->getData('certificates');
// $cert is complete, unfiltered list
```

Which results in the complete list, so no filtering is done.

```
dd($filter->toArray());
// Returns:
^ array:4 [▼
  "skip" => 0
  "limit" => 100
  "filter" => array:1 [▼
    0 => "common_names:["urliwanttoknowstatusof.nl"]"
  ]
  "sort" => []
]

dd(http_build_query($filter->toArray()));
// Returns:
skip=0&limit=100&filter%5B0%5D=common_names%3A%5B%22urliwanttoknowstatusof.nl%22%5D
```

I don't think the filter%5B0%5D parameter works, but I don't know how to set the filter without it.

**Comment by WilliamDEdwards** [comment]
FYI: we will be improving this behaviour in the API. `key` is currently searched for the literal `value`, which doesn't make too much sense when working with lists. Only the literal list, including all items in the list, is matched. However, we would expect `filter` to be mostly used to search for list items, instead of entire lists.

We will change/expand the behaviour of `filter` to either:

- Allow for searching for substring
- Literal match list items instead of only entire lists

Keep an eye out for the changelog.

**Comment by vdhicts** [comment]
@WilliamDEdwards could you please provide some information about how to use the filter? In the example of @marnickmenting it works in his Postman request (because he provides the `filter` parameter as string instead of an array) but can you show us how the request should look like when we want to provide multiple filter values?

**Comment by WilliamDEdwards** [comment]
> @WilliamDEdwards could you please provide some information about how to use the filter? In the example of @marnickmenting it works in his Postman request (because he provides the `filter` parameter as string instead of an array) but can you show us how the request should look like when we want to provide multiple filter values?
> 
> For example:
> 
> This works:
> 
> GET `/virtual-hosts?filter=domain:DOMAIN`
> 
> I've tried several things, but wasn't able to filter on multiple fields. Can you tell us how we are able to filter on domain and another field, for example "forceSsl"?

Specifying the `filter` query parameter multiple times should currently work.

**Comment by vdhicts** [comment]
When I perform:

GET `/virtual-hosts?filter=domain:DOMAIN&filter=forceSsl:false`

I'm still getting the virtualhost with the domain but the forceSsl property is true:

```
array(17) {
  ["domain"]=>
  string(14) "DOMAIN"
  ...
  ["force_ssl"]=>
  bool(true)
 ...
}
```

Is the filter an OR or AND?

**Comment by WilliamDEdwards** [comment]
> @WilliamDEdwards: when I perform:
> 
> GET `/virtual-hosts?filter=domain:DOMAIN&filter=forceSsl:false`
> 
> I'm still getting the virtualhost with the domain but the forceSsl property is true:
> 
> ```
> array(17) {
>   ["domain"]=>
>   string(14) "DOMAIN"
>   ...
>   ["force_ssl"]=>
>   bool(true)
>  ...
> }
> ```

The property is called `force_ssl`.

> 
> Is the filter an OR or AND?

AND

**Comment by vdhicts** [comment]
Ah thanks! I will fix the filter!

**Comment by vdhicts** [comment]
@WilliamDEdwards I've got it working now. For your information: one thing I noticed is that booleans needs to be represented by their numeric values. So `force_ssl:true` won't work but `force_ssl:1` does.

**Comment by vdhicts** [comment]
@marnickmenting you can expect a new release tomorrow in which this bug is fixed.

**Comment by vdhicts** [comment]
This problem is fixed in v1.24.1.

```
$filter = (new ListFilter())->setFilter(['domain:DOMAIN', 'force_ssl:1'])
$result = $api->virtualHosts()->list($filter); // Will return all virtual hosts with the domain AND force_ssl enabled
```

**Comment by WilliamDEdwards** [comment]
> @WilliamDEdwards I've got it working now. For your information: one thing I noticed is that booleans needs to be represented by their numeric values. So `force_ssl:true` won't work but `force_ssl:1` does.

Thanks, we will check and fix this.

**Comment by marnickmenting** [comment]
@WilliamDEdwards @vdhicts Thanks guys, it is working now :)

I use it for filtering certificates, it only matches if you provide the full array like this:

`$filter->setFilter(['common_names:["urliwanttoknowstatusof.nl"]']);`

With multiple common_names in one certificate it does not work. Which is fine for now, as I create the certificates one-by-one anyway.

@WilliamDEdwards knows about this, it should be changed in the API.

**Comment by vdhicts** [comment]
Thanks for your feedback, The filter is currently a literal match so that won't work indeed. As soon as the API is updated, I will update the package too.

**Comment by WilliamDEdwards** [comment]
> @WilliamDEdwards I've got it working now. For your information: one thing I noticed is that booleans needs to be represented by their numeric values. So `force_ssl:true` won't work but `force_ssl:1` does.

Fixed in API version 1.76! Version 1.76 will be released today.
